dot/core/service.go

```diff
@@ -543,12 +557,19 @@ func (s *Service) HandleSubmittedExtrinsic(ext types.Extrinsic) error {
 }
 
 // GetMetadata calls runtime Metadata_metadata function
-func (s *Service) GetMetadata(bhash *common.Hash) (metadata []byte, err error) {
-	rt, err := prepareRuntime(bhash, s.storageState, s.blockState)
+func (s *Service) GetMetadata(blockHash *common.Hash) (metadata []byte, err error) {
+	var blockHashValue common.Hash
+	if blockHash != nil {
+		blockHashValue = *blockHash
+	} else {
+		blockHashValue = s.blockState.BestBlockHash()
+	}
+	runtime, err := s.blockState.GetRuntime(blockHashValue)
 	if err != nil {
-		return nil, fmt.Errorf("setting up runtime: %w", err)
+		return nil, fmt.Errorf("getting runtime: %w", err)
 	}
-	return rt.Metadata()
+
+	return runtime.Metadata()
```

Comment:
same with this.. should we consider caching this?

Reply:
Yes, created https://github.com/ChainSafe/gossamer/issues/3065